src/python/pants/backend/code_quality_tool/lib.py

```diff
@@ -0,0 +1,291 @@
+from dataclasses import dataclass
+from typing import ClassVar, Type
+
+from pants.backend.adhoc.target_types import AdhocToolRunnableField, AdhocToolArgumentsField, \
+    AdhocToolRunnableDependenciesField, AdhocToolExecutionDependenciesField
+from pants.core.goals.fix import FixResult
+from pants.core.goals.fmt import FmtFilesRequest
+from pants.core.goals.lint import LintFilesRequest, LintResult
+from pants.core.goals.run import RunFieldSet, RunInSandboxRequest
+from pants.core.util_rules.adhoc_process_support import ResolvedExecutionDependencies, \
+    ResolveExecutionDependenciesRequest, rules as adhoc_process_support_rules, ExtraSandboxContents, \
+    MergeExtraSandboxContents
+from pants.core.util_rules.environments import EnvironmentNameRequest
+from pants.core.util_rules.partitions import Partitions, PartitionMetadata
+from pants.engine.addresses import Addresses
+from pants.engine.environment import EnvironmentName
+from pants.engine.fs import PathGlobs
+from pants.engine.internals.native_engine import FilespecMatcher, Snapshot, Address, AddressInput, Digest, EMPTY_DIGEST, \
+    MergeDigests
+from pants.engine.internals.selectors import Get
+from pants.engine.process import Process, FallibleProcessResult
+from pants.engine.rules import rule, collect_rules
+from pants.engine.target import Targets, FieldSetsPerTarget, FieldSetsPerTargetRequest, Target, COMMON_TARGET_FIELDS, \
+    StringSequenceField
+from pants.option.option_types import SkipOption
+from pants.option.subsystem import Subsystem
+from pants.util.frozendict import FrozenDict
+from pants.util.strutil import Simplifier
+
+
+class ByoFileGlobIncludeField(StringSequenceField):
+    alias: ClassVar[str] = "file_glob_include"
+    required = True
+
+
+class ByoFileGlobExcludeField(StringSequenceField):
+    alias: ClassVar[str] = "file_glob_exclude"
+    required = False
+    default = ()
+
+
+class CodeQualityToolTarget(Target):
+    alias: ClassVar[str] = "code_quality_tool"
+    core_fields = (
+        *COMMON_TARGET_FIELDS,
+        AdhocToolRunnableField,
+        AdhocToolArgumentsField,
+        AdhocToolExecutionDependenciesField,
+        AdhocToolRunnableDependenciesField,
+        ByoFileGlobIncludeField,
+        ByoFileGlobExcludeField,
+    )
+
+
+class ByoGoal:
+    goal: str
+    request_superclass: Type[LintFilesRequest]
+    skippable: list[str]
+    result_class: type
+
+    @classmethod
+    def create_result(cls, request, process_result, output):
+        raise NotImplementedError
+
+    @classmethod
+    def output_files_to_read(cls, request: LintFilesRequest.Batch):
+        raise NotImplementedError
```

Comment:
@lilatomic a ByoGoal is an internal (so probably should be `_ByoGoal`) object that factors out the differences between running the same tool but for different goals. it is a replacement for a bunch of `if/else`s that would be scattered around the rule definitions:
```
if goal == 'fmt':
    result_class = FmtResult
elif goal == 'lint':
    result_class = LintResult
```
and similarly for the request types, etc.

Reply:
Ah that makes sense, thanks.